subsys/shell/shell_mqtt.c

```diff
@@ -0,0 +1,775 @@
+/*
+ * Copyright (c) 2021 G-Technologies Sdn. Bhd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <shell/shell_mqtt.h>
+#include <init.h>
+#include <logging/log.h>
+#include <random/rand32.h>
+#include <string.h>
+#include <stdio.h>
+#include <drivers/hwinfo.h>
+
+SHELL_MQTT_DEFINE(shell_transport_mqtt);
+SHELL_DEFINE(shell_mqtt, NULL, &shell_transport_mqtt,
+	     CONFIG_SHELL_BACKEND_MQTT_LOG_MESSAGE_QUEUE_SIZE,
+	     CONFIG_SHELL_BACKEND_MQTT_LOG_MESSAGE_QUEUE_TIMEOUT, SHELL_FLAG_OLF_CRLF);
+
+LOG_MODULE_REGISTER(shell_mqtt, CONFIG_SHELL_MQTT_LOG_LEVEL);
+
+#define EVENT_MASK (NET_EVENT_L4_CONNECTED | NET_EVENT_L4_DISCONNECTED)
+#define APP_CONNECT_TIMEOUT_MS 2000
+#define APP_SLEEP_MSECS 500
+#define MQTT_SEND_DELAY_MS 100
+#define MQTT_PROCESS_INTERVAL 2
+#define SHELL_MQTT_WORKQ_STACK_SIZE 1024
+
+#ifdef CONFIG_SHELL_MQTT_SERVER_USERNAME
+#define MQTT_USERNAME CONFIG_SHELL_MQTT_SERVER_USERNAME
+#else
+#define MQTT_USERNAME NULL
+#endif /* CONFIG_SHELL_MQTT_SERVER_USERNAME */
+
+#ifdef CONFIG_SHELL_MQTT_SERVER_PASSWORD
+#define MQTT_PASSWORD CONFIG_SHELL_MQTT_SERVER_PASSWORD
+#else
+#define MQTT_PASSWORD NULL
+#endif /*SHELL_MQTT_SERVER_PASSWORD */
+
+struct shell_mqtt *sh_mqtt;
+K_THREAD_STACK_DEFINE(sh_mqtt_workq_stack, SHELL_MQTT_WORKQ_STACK_SIZE);
+
+static bool is_valid_char(const uint8_t *c)
+{
+	if (*c >= ' ' && *c <= '~') {
+		return true;
+	} else if (*c == '\r' || *c == '\n') {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+bool __weak shell_mqtt_get_devid(char *id, int id_max_len)
```

Comment:
Why is it weak?

Reply:
I hope that this can be overridden in application.

Reply:
Ok, but in such case I think it'd make sense to add a prototype in `shell_mqtt.h` header, along with some documentation explaining that this function can be redefined if needed.